modules/code-editor/src/backend/editor.ts

```diff
@@ -0,0 +1,130 @@
+import _ from 'lodash'
+import path from 'path'
+
+import { Config } from '../config'
+
+import { EditableFile, FileType, TypingDefinitions } from './typings'
+
+const FILENAME_REGEX = /^[0-9a-zA-Z_\-.]+$/
+
+export default class Editor {
+  private bp: typeof sdk
+  private _botId: string
+  private _typings: TypingDefinitions
+  private _config: Config
+
+  constructor(bp: typeof sdk, botId: string, config: Config) {
+    this.bp = bp
+    this._botId = botId
+    this._config = config
+  }
+
+  async fetchFiles() {
+    return {
+      actionsGlobal: this._config.allowGlobal && (await this._loadFiles('/actions', 'action')),
+      actionsBot: await this._loadFiles('/actions', 'action', this._botId)
+    }
+  }
+
+  async _validateContent({ name, botId, type }: Partial<EditableFile>) {
+    if (!botId || !botId.length) {
+      if (!this._config.allowGlobal) {
+        throw new Error(`Global files are restricted, please check your configuration`)
+      }
+    } else {
+      if (botId !== this._botId) {
+        throw new Error(`Please switch to the correct bot to change its actions.`)
+      }
+    }
+
+    if (type !== 'action') {
+      throw new Error('Invalid file type Only actions are allowed at the moment')
+    }
+
+    if (!FILENAME_REGEX.test(name)) {
+      throw new Error('Filename has invalid characters')
+    }
+  }
+
+  async saveFile(file: EditableFile): Promise<void> {
+    this._validateContent(file)
+    const { location, botId, content } = file
+    const ghost = botId ? this.bp.ghost.forBot(this._botId) : this.bp.ghost.forGlobal()
+
+    return ghost.upsertFile('/actions', location, content)
+  }
+
+  async loadTypings() {
+    if (this._typings) {
+      return this._typings
+    }
+
+    const sdkTyping = fs.readFileSync(path.join(__dirname, '/../botpress.d.js'), 'utf-8')
+
+    this._typings = {
+      'process.d.ts': this._buildRestrictedProcessVars(),
+      'node.d.ts': this._getNodeTypings().toString(),
+      'botpress.d.ts': sdkTyping.toString().replace(`'botpress/sdk'`, `sdk`)
+    }
+
+    return this._typings
+  }
+
+  private _getNodeTypings() {
+    const getTypingPath = folder => path.join(__dirname, `/../../${folder}/@types/node/index.d.ts`)
+
+    if (fs.existsSync(getTypingPath('node_modules'))) {
+      return fs.readFileSync(getTypingPath('node_modules'), 'utf-8')
+    }
+    return fs.readFileSync(getTypingPath('node_production_modules'), 'utf-8')
+  }
+
+  private async _loadFiles(rootFolder: string, type: FileType, botId?: string): Promise<EditableFile[]> {
+    const ghost = botId ? this.bp.ghost.forBot(botId) : this.bp.ghost.forGlobal()
+
+    return Promise.map(await ghost.directoryListing(rootFolder, '*.*'), async (filepath: string) => {
+      return {
+        name: path.basename(filepath),
+        location: filepath,
+        content: await ghost.readFileAsString(rootFolder, filepath),
+        type,
+        botId
+      }
+    })
+  }
+
+  private _buildRestrictedProcessVars() {
```

Comment:
Is there a way we get this from the core as I know we do the exact same ? I know it feels a bit odd to expose this in the sdk though ..

Reply:
I planned to do that in a future version, not sure on the best way to implement atm